### PR TITLE
Self-references and += operator

### LINKED ISSUE #35: support for "+=" field separator?
I get an error when parsing a hocon file:

```
[ec2-user@ip-172-16-2-117 ~]$ pyhocon -i /opt/mapr/drill/drill-1.1.0/conf/drill-override-example.conf
Traceback (most recent call last):
  File "/usr/local/bin/pyhocon", line 9, in <module>
    load_entry_point('pyhocon==0.3.4', 'console_scripts', 'pyhocon')()
  File "/usr/local/lib/python2.7/site-packages/pyhocon/tool.py", line 188, in main
    HOCONConverter.convert(args.input, args.output, args.format)
  File "/usr/local/lib/python2.7/site-packages/pyhocon/tool.py", line 145, in convert
    config = ConfigFactory.parse_file(input_file)
  File "/usr/local/lib/python2.7/site-packages/pyhocon/__init__.py", line 40, in parse_file
    return ConfigFactory.parse_string(content, os.path.dirname(filename))
  File "/usr/local/lib/python2.7/site-packages/pyhocon/__init__.py", line 75, in parse_string
    return ConfigParser().parse(content, basedir)
  File "/usr/local/lib/python2.7/site-packages/pyhocon/__init__.py", line 216, in parse
    config = config_expr.parseString(content, parseAll=True)[0]
  File "/usr/local/lib/python2.7/site-packages/pyparsing.py", line 1125, in parseString
    raise exc
pyparsing.ParseSyntaxException: Expected "{" (at char 1087), (line:20, col:33)
```

Line 20 in the above looks like this:

```
drill.logical.function.packages += "org.apache.drill.exec.expr.fn.impl"
```

And column 33 implicates the "+=" field separator.

Any plans to add support for this?
https://github.com/typesafehub/config/blob/master/HOCON.md#the--field-separator

### PR DESCRIPTION
**Pull request by gilles-duboscq**
This PR merges changes by @sanzinger that implement self-referential substitutions and the += operator.

The changes are a bit large so they are probably better reviewed commit by commit.

This would fix #35

### CODE REVIEW

**Comment by gilles-duboscq** [comment]
I'm not exactly sure what's going on with coveralls in the `py26` job.

**Comment by darthbear** [comment]
Thank you @gilles-duboscq ! This looks awesome! I'll review it this week.

File: pyhocon/config_parser.py [hunk 1]
```diff
@@ -291,40 +294,104 @@ def _resolve_variable(config, substitution):
...
+                    previous_item = current_item
+
+                if len(history) == 1: # special case, when self optional referencing without existing 
+                    for substitution in ConfigParser._find_substitutions(previous_item):
+                        prop_path = ConfigTree.parse_key(substitution.variable)
+                        if len(prop_path) > 1 and config.get(substitution.variable, None) is not None:
+                            continue # If value is present in latest version, don't do anything
+                        if prop_path[0] == key and substitution.optional:
+                            ConfigParser._do_substitute(substitution, None)
+
```
Comment (darthbear) [inline pyhocon/config_parser.py]: @sanzinger Really nice implementation for supporting self references! 👍

File: tests/test_config_parser.py [hunk 2]
```diff
@@ -706,6 +706,175 @@ def test_non_compatible_substitution(self):
...
+            x = ${x} {c: 3}
+            x = {z: 0} ${x}
+            x = {y: -1} ${x} {d: 4}
+            """
+        )
+        assert config.get("x") == {'a': 1, 'b': 2, 'c': 3, 'z': 0, 'y': -1, 'd': 4}
+
     def test_concat_multi_line_string(self):
         config = ConfigFactory.parse_string(
             """
```
Comment (darthbear) [inline tests/test_config_parser.py]: nice tests! 👍

File: pyhocon/config_parser.py [hunk 3]
```diff
@@ -494,8 +494,14 @@ def postParse(self, instring, loc, token_list):
...
-                    if isinstance(value, list):
-                        config_tree.put(key, value, operator == '+=')
+                    if isinstance(value, list) and operator == "+=":
+                        value = ConfigValues([ConfigSubstitution(key, True, '', False, loc), value], False, loc)
+                        config_tree.put(key, value, False)
+                    elif isinstance(value, str) and operator == "+=":
+                        value = ConfigValues([ConfigSubstitution(key, True, '', True, loc), ' ' + value], True, loc)
+                        config_tree.put(key, value, False)
+                    elif isinstance(value, list):
+                        config_tree.put(key, value, False)
```
Comment (darthbear) [inline pyhocon/config_parser.py]: @sanzinger Nice implementation of `+=` operator! 👍

**Comment by darthbear** [comment]
Thank you very much @sanzinger and @gilles-duboscq for implementing self reference and `+=` substitutions. Great pull request with elegant code! Just pushed it to pypi: `pyhocon==0.3.26` 🍻

**Comment by gilles-duboscq** [comment]
No problem.
All the credit goes to @sanzinger on this one. I just tried not to mess up the merge :wink:

Thanks for the quick integration.

**Comment by sanzinger** [comment]
Gilles, thank you for integrating that stuff. I never found the time to actually share it.

**Pull request merged**